src/nwb/NWBFile.hpp

```diff
@@ -121,6 +126,12 @@ class NWBFile
   void cacheSpecifications(const std::string& specPath,
                            const std::string& versionNumber);
 
+  /**
+   * @brief Whether the file is in recording mode (can write / append to
+   * existing datasets but cannot create new groups or datasets)
+   */
```

Comment:
Note, this is actually a HDF5-specific limitation. In Zarr this is not an issue because all objects are stored in separate directories. I think instead of using `isRecording` in the `NWBFIle` to determine if we can create new objects it would be best to ask the io backend if the operation is allowed or not, i.e., we could have a function in `BaseIO` (overwritten in `HDF5IO`) that says something like `canUpdateFileStructure()`. When SWMR mode is enable or we are in 'read' mode that function would return `False` and `True` otherwise. In this way we can avoid having HDF5-specific limitations trickle into the NWB API classes.

Reply:
I think this also means that the `isRecording` property should be on `BaseIO` and set it in `BaseIO.startRecording` and `HDF5IO.startRecording`

Reply:
> I think this also means that the isRecording property should be on BaseIO and set it in BaseIO.startRecording and HDF5IO.startRecording. 

I agree that makes sense, I've updated accordingly. 

> Note, this is actually a HDF5-specific limitation. In Zarr this is not an issue because all objects are stored in separate directories. I think instead of using isRecording in the NWBFIle to determine if we can create new objects it would be best to ask the io backend if the operation is allowed or not, i.e., we could have a function in BaseIO (overwritten in HDF5IO) that says something like canUpdateFileStructure(). When SWMR mode is enable or we are in 'read' mode that function would return False and True otherwise. In this way we can avoid having HDF5-specific limitations trickle into the NWB API classes.

I see what you are saying about having the HDF5-specific limitations trickle into the main classes. Currently the `isRecording` flag is really only being used to detect whether we can create new objects so I replaced in entirely with `canModifyObjects`. 

I think if we establish a general usage pattern of 1. setup file structure, 2. create all datasets 3. record to datasets 4. stop recording and perform final edits / cleanup, it could be simpler to keep the name as `isRecording`? `createFileStructure` and `create{X}Series` are the main functions that currently need to consider this property and at least `createFileStructure` should probably not be called mid recording? But if `canModifyObjects` or some naming like that is clearer as to the intention, and if adding new objects during the recording process is something that Zarr / non-SWMR mode HDF5 files would want to do, then that works.

Reply:
> I see what you are saying about having the HDF5-specific limitations trickle into the main classes. Currently the `isRecording` flag is really only being used to detect whether we can create new objects so I replaced in entirely with `canModifyObjects`.

That looks good to me. 

> I think if we establish a general usage pattern of 1. setup file structure, 2. create all datasets 3. record to datasets 4. stop recording and perform final edits / cleanup, it could be simpler to keep the name as `isRecording`?

For now, I'm leaning towards keeping it as `canModifyObjects`, mainly because this is really just a restriction of `SWMR` mode and will likely not be needed by backends like Zarr. I agree that we want to promote this pattern as a best practice, but I'm not sure right now if want to make it a hard restriction yet for all backends.